File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.py
# 프로그래머스 84021 퍼즐 조각 채우기 LV.3
from collections import deque

dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]


def bfs(begin, target, return_list, r_c_len, original, changed):
    i, j = begin
    q = deque()
    q.append(begin)
    target[i][j] = changed

    while q:
        x, y = q.popleft()
        return_list.append([x, y])
        for d in range(4):
            nx, ny = x + dx[d], y + dy[d]
            if 0 <= nx < r_c_len and 0 <= ny < r_c_len and target[nx][ny] == original:
                q.append([nx, ny])
                target[nx][ny] = changed


def arrange_relative_coordinates(target):
    target.sort()
    x0, y0 = target[0]

    for k in range(len(target)):
        target[k][0] = target[k][0] - x0
        target[k][1] = target[k][1] - y0


def solution(game_board, table):
    answer = 0
    blocks = []
    m = len(game_board)

    # 블록 찾기 -> 찾아서 blocks 에 좌표 저장 (시작이 (0, 0) 으로 저장)
    for i in range(m):
        for j in range(m):
            if table[i][j] == 1:
                # bfs 로 block 찾기
                block = []
                bfs([i, j], table, block, m, 1, 0)

                # block 의 좌표를 상대 좌표로 정리하기
                arrange_relative_coordinates(block)

                # blocks 에 저장
                blocks.append(block)

    # 빈칸 찾기 -> 찾아서 blanks 에 좌표 저장(상대 좌표) -> block 과 blank 일치하는지 체크
    for i in range(m):
        for j in range(m):
            if game_board[i][j] == 0:
                blank = []
                bfs([i, j], game_board, blank, m, 0, 1)

                # blank 의 좌표를 상대 좌표로 정리하기
                arrange_relative_coordinates(blank)

                # 현재 blank 와 일치하는 block 찾기 -> 회전하면서 일치하면 정답 기록
                for b in range(len(blocks)):
                    cur_block = blocks[b]
                    flag = False

                    # block 시계 방향 회전 (90 -> 180 -> 270 -> 360)
                    for c in range(4):
                        for idx in range(len(cur_block)):
                            cur_block[idx][0], cur_block[idx][1] = cur_block[idx][1], -cur_block[idx][0]

                        # 상대 좌표 정리
                        arrange_relative_coordinates(cur_block)

                        # 일치하면 해당 블록 제거
                        if blank == cur_block:
                            del blocks[b]
                            answer += len(cur_block)
                            flag = True
                            break

                    if flag:
                        break

    return answer

File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.java
import java.util.*;

class Solution {
    int[] dy = new int[]{-1, 0, 1, 0}, dx = new int[]{0, 1, 0, -1};

    public int solution(int[][] game_board, int[][] table) {
        int answer = 0;

        // 1. 블록 좌표 찾기 -> 블록 상대좌표로 만들어서 저장하기
        List<List<int[]>> blocks = new ArrayList<>();

        for (int i = 0; i < table.length; i++) {
            for (int j = 0; j < table.length; j++) {
                if (table[i][j] == 1) {
                    List<int[]> block = new ArrayList<>();

                    bfs(table, block, new int[]{i, j}, 1, 0);

                    // 상대 좌표로 저장
                    convertRelativeCoord(block);

                    // 블록 저장
                    blocks.add(block);
                }
            }
        }

        // 2. 빈칸 찾기 -> 빈칸 찾아서 블록의 상대좌표랑 일치하는지 확인 -> 일치하면 블록 제거하고 답 기록
        for (int i = 0; i < game_board.length; i++) {
            for (int j = 0; j < game_board.length; j++) {
                if (game_board[i][j] == 0) {
                    List<int[]> blank = new ArrayList<>();

                    bfs(game_board, blank, new int[]{i, j}, 0, 1);

                    convertRelativeCoord(blank);

                    for (int b = 0; b < blocks.size(); b++) {
                        List<int[]> currentBlock = blocks.get(b);
                        boolean flag = false;

                        for (int k = 0; k < 4; k++) {
                            for (int[] cb : currentBlock) {
                                int tmp = cb[0];
                                cb[0] = cb[1];
                                cb[1] = -tmp;
                            }

                            convertRelativeCoord(currentBlock);

                            if (matchesBlankAndBlock(blank, currentBlock)) {
                                blocks.remove(b);
                                answer += currentBlock.size();
                                flag = true;
                                break;
                            }
                        }

                        if (flag) {
                            break;
                        }
                    }
                }
            }
        }

        return answer;
    }

    public void bfs(int[][] matrix, List<int[]> result, int[] coord, int original, int changed) {
        Queue<int[]> q = new LinkedList<>();
        int y = coord[0], x = coord[1];
        q.offer(coord);
        matrix[y][x] = changed;

        while (!q.isEmpty()) {
            int[] current = q.poll();
            int cy = current[0], cx = current[1];

            result.add(new int[]{cy, cx});

            for (int i = 0; i < 4; i++) {
                int ny = cy + dy[i], nx = cx + dx[i];

                // 범위 안에 존재하고
                if (0 <= ny && ny < matrix.length && 0 <= nx && nx < matrix.length) {
                    // 찾은 좌표의 값이 original 이라면
                    if (matrix[ny][nx] == original) {
                        q.offer(new int[]{ny, nx});
                        matrix[ny][nx] = changed;
                    }
                }
            }
        }
    }

    public void convertRelativeCoord(List<int[]> coords) {
        coords.sort((o1, o2) -> {
            if (o1[0] == o2[0]) {
                return o1[1] - o2[1];
            }
            return o1[0] - o2[0];
        });
        int y0 = coords.get(0)[0], x0 = coords.get(0)[1];

        for (int[] c : coords) {
            c[0] -= y0;
            c[1] -= x0;
        }
    }

    public boolean matchesBlankAndBlock(List<int[]> blank, List<int[]> block) {
        if (blank.size() != block.size()) {
            return false;
        }

        for (int i = 0; i < blank.size(); i++) {
            if (!Arrays.equals(blank.get(i), block.get(i))) {
                return false;
            }
        }

        return true;
    }
}